torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -25,6 +26,97 @@
 from torch.distributed.utils import _to_kwargs
 
 
+@no_type_check
+def _lazy_init(
+    state: _State,
+    root_module: nn.Module,
+) -> _State:
+    """
+    Performs initialization lazily, typically right before the first forward
+    pass. The laziness is needed to ensure that the parameter device/dtype and
+    the FSDP hierarchy have finalized. This method's actual logic only runs on
+    the root FSDP instance, which performs initialization for all non-root FSDP
+    instances to avoid partial initialization.
+
+    For the non-composable code path, ``state`` and ``root_module`` should be
+    the same, namely the FSDP instance itself.
+    """
+    if state._is_root is not None:
+        return  # no-op: already lazily initialized
+    if not torch.cuda.is_available():
+        # Allow the FSDP constructor to run even without CUDA but check this
+        # once we start real execution
+        raise RuntimeError("FSDP does not support CPU only execution")
+    # The following logic is only run on the root FSDP instance since it will
+    # set `_is_root=False` for the non-root instances
+    state._is_root = True
+    _assert_in_training_states(state, [TrainingState.IDLE])
+    _init_streams(state)
+    buffers, buffer_dtypes = _get_buffers_and_dtypes_for_computation(state, root_module)
+    _cast_buffers_to_dtype_and_device(buffers, buffer_dtypes, state.compute_device)
+    for handle in state._handles:
+        handle.init_flat_param_attributes()
+    state._exec_order_data.init(state, state.process_group)
+    if _is_composable(state):
+        # Return early since there is no need to share data structures
+        return state
```

Comment:
I thought composable FSDP still need a local root, so that it can manage its children instances and make sure they could have overlapped communication & computation?

Reply:
The module passed into the composable FSDP API is the local root. Hooks are partially applied with references to the `state: _State` object, which gives them access to the shared data structures.

The current implementation cannot handle sharing across separate applications of `fully_shard()`.